dotnet-statsig/src/Statsig/Server/Evaluation/SpecStore.cs

```diff
@@ -24,104 +25,199 @@ internal IDList(string name)
 
     class SpecStore
     {
-        double _lastSyncTime;
-        RequestDispatcher _requestDispatcher;
-        Timer _syncTimer;
-        Timer _idListSyncTimer;
+        private double _lastSyncTime;
+        private readonly RequestDispatcher _requestDispatcher;
 
-        internal Dictionary<string, ConfigSpec> FeatureGates { get; set; }
-        internal Dictionary<string, ConfigSpec> DynamicConfigs { get; set; }
-        internal Dictionary<string, IDList> IDLists { get; set; }
+        private Task _syncIDListsTask;
+        private Task _syncValuesTask;
+        private readonly CancellationTokenSource _cts;
+
+        internal Dictionary<string, ConfigSpec> FeatureGates { get; private set; }
+        internal Dictionary<string, ConfigSpec> DynamicConfigs { get; private set; }
+        private readonly Dictionary<string, IDList> _idLists;
+        private readonly ReaderWriterLockSlim _idListLock;
 
         internal SpecStore(string serverSecret, StatsigOptions options)
         {
             _requestDispatcher = new RequestDispatcher(serverSecret, options.ApiUrlBase);
             _lastSyncTime = 0;
             FeatureGates = new Dictionary<string, ConfigSpec>();
             DynamicConfigs = new Dictionary<string, ConfigSpec>();
-            IDLists = new Dictionary<string, IDList>();
+            _idLists = new Dictionary<string, IDList>();
+            _idListLock = new ReaderWriterLockSlim();
+
+            _syncIDListsTask = null;
+            _syncValuesTask = null;
+            _cts = new CancellationTokenSource();
         }
 
         internal async Task Initialize()
         {
+            // Execute and wait for the initial syncs
             await SyncValues(true);
             await SyncIDLists();
+
+            // Start background tasks to periodically refresh the store
+            _syncValuesTask = BackgroundPeriodicSyncValuesTask(_cts.Token);
+            _syncIDListsTask = BackgroundPeriodicSyncIDListsTask(_cts.Token);
         }
 
-        internal void Shutdown()
+        internal async Task Shutdown()
         {
-            _syncTimer.Stop();
-            _syncTimer.Dispose();
-            _idListSyncTimer.Stop();
-            _idListSyncTimer.Dispose();
+            // Signal that the periodic task should exit, and then wait for them to finish
+            _cts.Cancel();
+            if (_syncIDListsTask != null)
+            {
+                await _syncIDListsTask;
+            }
+            if (_syncValuesTask != null)
+            {
+                await _syncValuesTask;
+            }
+        }
+
+        internal bool IDListContainsValue(string listName, string value)
+        {
+            _idListLock.EnterReadLock();
+            try
+            {
+                return _idLists.TryGetValue(listName, out var list) && list.IDs.Contains(value);
+            }
+            finally
+            {
+                _idListLock.ExitReadLock();
+            }
```

Comment:
curious how performant is this if QPS is in the thousands?

Reply:
It _should_ be fine.  If you're curious about the implementation, you can read it [here](https://source.dot.net/#System.Private.CoreLib/ReaderWriterLockSlim.cs,08a579fe90a54f7a).

Another option here could be to do a wholesale replacement of the _idLists whenever it's being updated, rather than trying to modify it in-place.  Then we could get rid of the locking on the read path, and only do locking on the update path to ensure we don't have concurrent/overlapping updates.

Reply:
Yeah, that's something we are doing with the `download_config_specs` periodical updates. Tradeoff would be that we can potentially be holding 2 copies of id lists in memory whenever this runs, and it could increase memory usage by quite a bit if a company uses a lot of large lists. Performance is something we can optimize later though, so I think the tradeoff is ok at least for now